pkg/controllers/manifestwork.go

```diff
@@ -0,0 +1,250 @@
+			Annotations: map[string]string{
+				CreatedByHypershiftDeployment: fmt.Sprintf("%s%s%s",
+					hyd.GetNamespace(),
+					NamespaceNameSeperator,
+					hyd.GetName()),
+			},
+		},
+		Spec: workv1.ManifestWorkSpec{},
+	}
+}
+
+func getManifestWorkKey(hyd *hypdeployment.HypershiftDeployment) types.NamespacedName {
+	return types.NamespacedName{
+		Name:      hyd.GetName(),
+		Namespace: getTargetNamespace(hyd),
+	}
+}
+
+func (r *HypershiftDeploymentReconciler) createMainfestwork(ctx context.Context, req ctrl.Request, hyd *hypdeployment.HypershiftDeployment) (ctrl.Result, error) {
+	m := ScafoldManifestwork(hyd)
+
+	// if the manifestwork is created, then move the status to hypershiftDeployment
+	// TODO: @ianzhang366 might want to do some upate/patch when the manifestwork is created.
+	if err := r.Get(ctx, getManifestWorkKey(hyd), m); err == nil {
+		workConds := m.Status.Conditions
+
+		for _, cond := range workConds {
+			if err := r.updateStatusConditionsOnChange(
+				hyd,
+				hypdeployment.ConditionType(cond.Type),
+				metav1.ConditionTrue,
+				cond.Message,
+				cond.Reason,
+			); err != nil {
+				r.Log.Info(fmt.Sprintf("update status condition failed for %s%s, err: %v", getTargetNamespace(hyd), hyd.GetName(), err))
+				return ctrl.Result{RequeueAfter: 10 * time.Second, Requeue: true}, nil
+			}
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	appendSecrets, err := r.appendReferenceSecrets(ctx, hyd)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	payload := []workv1.Manifest{}
+
+	manifestFuncs := []loadManifest{
+		appendHostedCluster,
+		appendNodePool,
+		appendSecrets,
+	}
+
+	for _, f := range manifestFuncs {
+		f(hyd, &payload)
+	}
+
+	m.Spec.Workload.Manifests = payload
+
+	// 	if err := controllerutil.SetOwnerReference(hyd, m, r.Scheme); err != nil {
+	// 		return ctrl.Result{}, fmt.Errorf("failed to set manifestwork's owner as %s, err: %w", req, err)
+	// 	}
+
+	if err := r.Create(r.ctx, m); err != nil {
+		if apierrors.IsAlreadyExists(err) {
+			//TODO: ianzhang366, might want to patch the manifestwork over here.
+			return ctrl.Result{}, r.Update(r.ctx, m)
+		}
+
+		return ctrl.Result{}, fmt.Errorf("failed to create manifestwork based on hypershiftDeployment: %s, err: %w", req, err)
+	}
+
+	r.Log.Info(fmt.Sprintf("created manifestwork for hypershiftDeployment: %s at targetNamespace: %s", req, getTargetNamespace(hyd)))
+
+	return ctrl.Result{}, nil
+}
+
+func (r *HypershiftDeploymentReconciler) deleteManifestworkWaitCleanUp(ctx context.Context, hyd *hypdeployment.HypershiftDeployment) (ctrl.Result, error) {
+
+	m := ScafoldManifestwork(hyd)
+
+	if err := r.Delete(ctx, m); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return ctrl.Result{}, fmt.Errorf("failed to delete manifestwork, err: %v", err)
+		}
+	}
+
+	for _, np := range hyd.Spec.NodePools {
+		var nodePool hyp.NodePool
+		if err := r.Get(ctx, types.NamespacedName{Namespace: getTargetNamespace(hyd), Name: np.Name}, &nodePool); err == nil {
```

Comment:
Hi @ianzhang366. @jnpacker pointed me to this pull request and I am trying to understand how this would work, in case we want to use this mechanism for OCM. I apologize in advance if I ask dumb questions.

Is this expecting the `NodePool` and `HostedCluster` objects to exist on the same cluster where the `ManifestWork` exists? Won't those have been created on the target cluster of the `ManifestWork` instead? Also, doesn't the manifest work controller on the target cluster take care of deleting and waiting for clean up in the target cluster?

Reply:
yeah, I think jhernand is right, in the case of one managed cluster as the hypershift management cluster, get the HostedCluster and NodePool will fail(since they do not exist on the hub cluster).

Reply:
yes... when put this up, I was having the `local-cluster` mode in mind.

the idea was, the hub will create the infrastructure and then render the `nodepool` and `hostedcluster` for the target managed cluster to use.

If we delete the `manifestwork` then delete the infra right away, the managedcluster's hypershift operator might unable to clean up the `nodepool` and `hostedcluster` correctly... @jnpacker do you know details for this?

another way to do this would be, we ship the hypershiftDeployment controller to the managedcluster, and on hub, we wrap the `hypershiftDeployment` and the provider's secret to a manifestwork.

Reply:
another update.

after some test, it seems the `manifestwork` has a finalizer that will be removed till all created resources are deleted.

Given these, now we will wait for the `manifestwork` to be deleted instead of watching the actual `hostedcluster`, 'nodepool' cr.